keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,552 @@
+	// +optional
+	SuccessfulJobsHistoryLimit *int32
+
+	// The number of failed finished jobs to retain.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// +optional
+	FailedJobsHistoryLimit *int32
+}
+
+// ConcurrencyPolicy describes how the job will be handled.
+// Only one of the following concurrent policies may be specified.
+// If none of the following policies is specified, the default one
+// is AllowConcurrent.
+type ConcurrencyPolicy string
+
+const (
+	// AllowConcurrent allows CronJobs to run concurrently.
+	AllowConcurrent ConcurrencyPolicy = "Allow"
+
+	// ForbidConcurrent forbids concurrent runs, skipping next run if previous
+	// hasn't finished yet.
+	ForbidConcurrent ConcurrencyPolicy = "Forbid"
+
+	// ReplaceConcurrent cancels currently running job and replaces it with a new one.
+	ReplaceConcurrent ConcurrencyPolicy = "Replace"
+)
+
+// CronJobConditionType for CronJob conditions
+type CronJobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// CronJobSettled means the cron job controller has
+	CronJobSettled JobConditionType = "Settled"
+)
+
+// CronJobCondition describes a condition state
+type CronJobCondition struct {
+	// Type of job condition, Settled
+	Type CronJobConditionType
+	// Status of the condition, one of True, False, Unknown.
+	Status api.ConditionStatus
+	// Last time the condition was checked.
+	// +optional
+	LastProbeTime metav1.Time
+	// Last time the condition transit from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time
+	// (brief) reason for the condition's last transition.
+	// +optional
+	Reason string
+	// Human readable message indicating details about last transition.
+	// +optional
+	Message string
+}
+
+// CronJobStatus represents the current state of a cron job.
+type CronJobStatus struct {
+	// A list of pointers to currently running jobs.
+	// +optional
+	Active []api.ObjectReference
+
+	// Information when was the last time the job was successfully scheduled.
+	// +optional
+	LastScheduleTime *metav1.Time
+
+	// Information for when was the next time the job will be scheduled.
+    // May not be accurate in some cases for Forbid concurrency policy
+	// +optional
+	NextScheduleTime *metav1.Time
+
+	// Information when was the last time the job successfully completed.
+	// +optional
+	LastSuccessfulTime *metav1.Time
+
+	// Counter for tracking successful Job runs
+	// +optional
+	SuccessfulRuns int64
+
+	// Counter for tracking failed Job runs
+	// +optional
+    FailedlRuns int64
+
+	// Counter for tracking failed Job runs since last successful run
+	// +optional
+    FailuresSinceSuccess int64
+
+	// The set of conditions for this object
+	// +optional
+	Conditions []CronJobCondition
+
+}
+```
+
+## v1beta1 changes
+None
+
+## Validations
+Nothing additional from v1beta1
+
+## Tests
+
+### E2E test
+CronJob E2E test code is [located here](`/test/e2e/apps/cronjob.go`)
```

Comment:
Make the link workable: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/cronjob.go